[QoI] [Parse] Improve error message when parsing integer literal

d-ronnqvist opened the pull request:
<!-- What's in this pull request? -->
This PR does a few things to improve the clarity of the diagnostics when parsing/lexing integer literals:

1. It rephrases an error message to say "in integer literal" instead of "after integer literal prefix"
2. It provides additional information about what digits are expected (binary, octal, decimal, hexadecimal)
3. It looks ahead once a integer literal has been parsed to check for unexpected digits.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-5263](https://bugs.swift.org/browse/SR-5263).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

d-ronnqvist committed:
[QoI] Improve error message when parsing integer literal

Rephrased error message to indicate which character is unexpected.
Provide error message variations when parsing binary, octal, decimal (default), and hexadecimal integer literals.
Look for unexpected digits in binary and octal integer literals.
Look for unexpected letters in hex integer literals.

Resolves: SR-5236 rdar://problem/32858684

d-ronnqvist commented:
I think this is an improvement, but I'm not sure if it's the right solution or if it's too minor.

Unless the "prefix" part is an important distinction in the existing error message, points 1 and 2 should be pretty straight forward. However, point 3 results in a change of behavior. For example, `0b1020` is currently lexed as two integer literals (`0b10` and `20`) immediately following each other which would result in an error about separating statements with semicolon:

    error: consecutive statements on a line must be separated by ';'
    0b1020
        ^
        ;


This now forms an unknown token (similar to how `0b2` is currently treated) with a diagnostic about an unexpected digit:

    error: expected a binary digit (0-1) in integer literal
    0b1020
        ^

If I'm correct that it's never valid to have two literals after each other, I don't think this would cause any regressions (and that it provides a clearer error message).

A similar lookahead exists for octal literals and hex literals

    error: expected an octal digit (0-7) in integer literal
    0o7080
        ^

    error: expected a hex digit (0-9,A-F) in integer literal
    0xF0G0
        ^

I find that the binary literal and octal literal checks provide a helpful diagnostic for developers who may add other digits in those literals. I think that it's good to have hex literals provide the same diagnostic, but I'm not sure if it's as common for developers to include other letters in hex literals.

d-ronnqvist commented:
Also, `git-clang-format` wants to format 

    while (advanceIfValidContinuationOfIdentifier(CurPtr, BufferEnd));

as

    while (advanceIfValidContinuationOfIdentifier(CurPtr, BufferEnd))
        ;

but it wasn't previously formatted like that and doesn't seem to be formatter like that elsewhere in this file, so I discarded that specific change.

huonw commented:
Those error messages do seem much better! (Minor point: I think you flipped the last two digits in the SR, it should be https://bugs.swift.org/browse/SR-5263 .)

huonw commented on lib/Parse/Lexer.cpp:
```diff
@@ -932,20 +932,27 @@ void Lexer::lexHexNumber() {
-  if (!isHexDigit(*CurPtr)) {
-    diagnose(CurPtr, diag::lex_expected_digit_in_int_literal);
+  auto expected_hex_digit = [&]() {
+    diagnose(CurPtr, diag::lex_expected_digit_in_int_literal, 3 /*Hex*/);
```
The code might be slightly neater/more self-documenting with an `enum` like,

```cpp
enum class ExpectedDigitKind : unsigned {
    Binary,
    Octal, 
    Decimal,
    Hex
}
```

which would mean these calls could look like `diagnose(..., (unsigned)ExpectedDigitKind::Hex)` here and `expected_int_digit(..., ExpectedDigitKind::Binary)` below (after adjusting the closure to take `ExpectedDigitKind`).

(If you're interested in an example to crib off, the `where_without_generic_params` diagnostic does something similar with the `Parser::WhereClauseKind` enum.)

d-ronnqvist replied:
I think it would have to be available in all of the lexer since it would be used from two methods (`lexNumber` and `lexHexNumber`). Is that still preferable even if it's only used for a diagnostic?

huonw replied:
Yes, I think so, although it should only need to be available in the bit of this file starting from above this function, and all the uses should be fairly close together.

huonw commented on lib/Parse/Lexer.cpp:
```diff
@@ -1013,34 +1020,47 @@ void Lexer::lexHexNumber() {
     while ((*CurPtr >= '0' && *CurPtr <= '7') || *CurPtr == '_')
       ++CurPtr;
+
+    if (isHexDigit(*CurPtr))
```
I'm curious why this is testing for a hex digit; maybe alphanumeric would work and cover more cases?

rintaro replied:
I think `if (advanceIfValidContinuationOfOperator(CurPtr, BufferEnd))` is better.

jrose-apple replied:
Why operator?

rintaro replied:
Ah thanks, my bad, `advanceIfValidContinuationOfIdentifier` of course.

d-ronnqvist replied:
It seems like `isValidIdentifierContinuationCodePoint` does the same check as `advanceIfValidContinuationOfIdentifier` but without advancing `CurPtr`.

rintaro replied:
You can't directly use `isValidIdentifierContinuationCodePoint` because it receives `uint32_t ` Unicode code point, not UTF8 code point.
`advanceIfValidContinuationOfIdentifier` decodes UTF8 sequence to Unicode code point.

For example, we want to tokenize `12号`  as single `tok::unkown`, no?

If you don't want to advance `CurPtr`, you could use a temporary pointer for this:

```cpp
auto tmpPtr  = CurPtr;
if (advanceIfValidContinuationOfIdentifier(tmpPtr, BufferEnd)) { ...
```

However, in most cases, you `should` consume it because `expected_digit` consumes them anyway.

d-ronnqvist replied:
Ah, I didn't think of that. I'll change them all. 
@rintaro Should I be checking `advanceIfValidStartOfIdentifier` (note `Start`) for the diagnostic relating to the first digit of an exponent? Or should that still be `advanceIfValidContinuationOfIdentifier` (note `Continuation`)?

rintaro replied:
I'd go with `advanceIfValidContinuationOfIdentifier`.
`1e$0` is invalid and can be tokenized as single unknown.
(`$` is valid "continuation", but not "start")

jrose-apple commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -150,7 +150,8 @@ ERROR(lex_invalid_identifier_start_character,none,
        "expected a digit in floating point exponent", ())
 ERROR(lex_expected_digit_in_int_literal,none,
-       "expected a digit after integer literal prefix", ())
+      "expected %select{a binary digit (0-1)|an octal digit (0-7)|a digit|a hex"
```
Nitpick: let's say `0 or 1` for binary digit.

xwu replied:
Another nit: since the other messages (see immediately below) also spell it out, probably more consistent to say "hexadecimal digit" instead of "hex digit"; also, could we put a space in `(0-7,A-F)` after the comma?

jrose-apple reviewed (commented):
I like this too, and it's much better than parsing it as two tokens. I think we could go even further and say `'G' is not a valid hexadecimal digit (0-9, A-F) in integer literal`, i.e. actually point out which character is the problem, but that's just icing on the cake.

rintaro commented:
Please cover identifier characters after exponent part as well, e.g. `0xffp+12abc` `12e1abc`.

d-ronnqvist committed:
[QoI] [Parse] Improve error message when parsing floating point exponent

Update error messages to mention the invalid character.
Improve the diagnostic of floating point exponents.

Add tests for error messages when parsing floating point exponents.
Update existing tests for new error messages.

d-ronnqvist commented:
I wasn't entirely satisfied with how the new diagnostic applied to some exponent cases, so I both kept the current error message and added the new error message. For example: 

 * `1e+` would have resulted in `' ' is not a valid digit in floating point exponent`. Here the current message makes more sense (`expected a digit in floating point exponent`) since the exponent is essentially empty.
 * `1e_` would have resulted in `'\_' is not a valid digit in floating point exponent`. This makes it sound like `_` is not allowed in exponents, but that only applies to the first character. Here a variation of the diagnostic is used to indicate this distinction: `'\_' is not a valid first character in floating point exponent`.

In other cases the new error message works well. For example:

 * `0xffp+12abc` produces `'a' is not a valid digit in floating point exponent`.
 * `12e1abc` produces `'a' is not a valid digit in floating point exponent`.

d-ronnqvist committed:
Fix unicode handling when checking invalid characters

Use `advanceIfValidContinuationOfIdentifier` instead of `isValidIdentifierContinuationCodePoint` to handle unicode.

jrose-apple commented on test/Parse/number_identifier_errors.swift:
```diff
@@ -36,23 +36,23 @@ typealias 11.0 = Int
 enum 18enum {}
 // expected-error@-1 {{enum name can only start with a letter or underscore, not a number}}
-// expected-error@-2 {{expected a digit in floating point exponent}}
+// expected-error@-2 {{'n' is not a valid digit in floating point exponent}}
```
Yuck. Follow-up work to see if the floating-point case doesn't have a fractional part and therefore is more likely to be a malformed identifier? :-) (Not your fault and not a regression, so would not have to be part of this PR.)

huonw replied:
This is, I guess, covered by rdar://problem/32901394 (I'm not sure if there's a JIRA for it).

jrose-apple replied:
(For David's benefit, there's nothing particularly interesting in that Radar: "Need better compiler diagnostics when starting a function declaration with a number".)

huonw reviewed (approved):
Looks great! Thanks for iterating on it.

huonw commented:
@swift-ci Please smoke test
